keras_nlp/models/gpt_neo_x/gpt_neo_x_attention.py

```diff
@@ -0,0 +1,152 @@
+# Copyright 2023 The KerasNLP Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.models.gpt_neo_x.rotary_embedding import RotaryEmbedding
+from keras_nlp.utils.keras_utils import clone_initializer
+
+
+class GPTNeoXAttention(keras.layers.Layer):
+    def __init__(
+        self,
+        num_heads,
+        hidden_dim,
+        dropout=0.1,
+        max_sequence_length=512,
+        kernel_initializer="glorot_uniform",
+        bias_initializer="zeros",
+        rotary_percentage=0.25,
+        rotary_max_wavelength=10000,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.num_heads = num_heads
+        self.hidden_dim = hidden_dim
+        self.rotary_percentage = rotary_percentage
+        self.dropout = dropout
+        self.attn_head_size = hidden_dim // num_heads
+        self.rotary_ndims = int(self.attn_head_size * rotary_percentage)
```

Comment:
It seems to me like you can move this line down into the rotary layer itself, you can get at `attn_head_size` simply by reading the shape of the passed `query` and `value` right?

I would pass `percentage` and `max_wavelength` directly as arguments to `RotaryEmbedding`, and keep all the logic there, that will keep things more compartmentalized.

Reply:
outputs match when doing this refactor :)